### ISSUE TITLE
BulkheadAsynchRetryTest.testRetriesJoinBackOfQueue seems wrong

### ISSUE THREAD

**Issue by Ladicek** [eclipse/microprofile-fault-tolerance#409]
My reading of the test `BulkheadAsynchRetryTest.testRetriesJoinBackOfQueue` suggests that it's implementation-dependent. That is, I believe it's possible to write an compliant implementation that fails this test. I'm specifically thinking:

1. The beginning of the test reads:

```java
AsyncBulkheadTask taskA = newTask();
Future resultA = retryQueueAsyncBean.test(taskA);

AsyncBulkheadTask taskB = newTask();
Future resultB = retryQueueAsyncBean.test(taskB);

AsyncBulkheadTask taskC = newTask();
Future resultC = retryQueueAsyncBean.test(taskC);

taskA.assertStarting(resultA);
taskB.assertNotStarting();
taskC.assertNotStarting();
```

This test expects the first call will be the first to make it into the bulkhead, while there's nothing in the test that would affect thread ordering. Clearly this presumes one implementation strategy, while other implementation strategies are also valid.

2. The test javadoc also says

> Test that when an execution is retried, it goes to the back of the bulkhead queue.

This itself again presumes an implementation strategy, while there are other valid implementation strategies.

WDYT?

**Comment by Azquelt** [comment]
1. Yes, I guess if an implementation was starting a new thread for each task which would then attempt to acquire a slot on the bulkhead, then they could happen in the wrong order.

Would it satisfy you if that code was changed to this?
```
        AsyncBulkheadTask taskA = newTask();
        Future resultA = retryQueueAsyncBean.test(taskA);
        taskA.assertStarting(resultA);
        
        AsyncBulkheadTask taskB = newTask();
        Future resultB = retryQueueAsyncBean.test(taskB);
        taskB.assertNotStarting();
        
        AsyncBulkheadTask taskC = newTask();
        Future resultC = retryQueueAsyncBean.test(taskC);
        taskC.assertNotStarting();
```

2. I'm not sure about this. The spec says that an asynchronous bulkhead has a waiting task queue, which to me implies a first-in-first-out operation, though I'd welcome other opinions on this. It also says that when an execution throws an exception, it leaves the bulkhead, waits for the configured delay, and then tries to re-enter it again.

Therefore, when taskA throws an exception, I'd expect it to leave the bulkhead, re-enter at the back of the queue, for taskB to run next, and when that finishes, I'd expect taskC to run.

**Comment by Ladicek** [comment]
Re 1: That change would certainly improve the test, but there's subsequent code that again relies on ordering without justification:

```java
taskA.completeExceptionally(new TestException()); // fail A, causes instant retry which puts taskA on the back of the queue

taskB.assertStarting(resultB);
taskC.assertNotStarting();
```

Not sure if there's an easy fix like for the previous code.

Re 2: You're right the spec says that, but I believe the spec went too far with this. If the retry has no delay and no jitter (just like the test in question), I believe it's very much valid choice to retry immediately without leaving the bulkhead.

**Comment by Azquelt** [comment]
Re 1: if the waiting task queue is a first-in-first-out queue, then taskB should always run at this point (rather than taskC) because it was added to the queue first.

Re 2: That's fair comment, but you're then asking for a change to the design of the spec rather than just a fix to one of the tests. It might be better to raise a different issue for that one as it'll need a longer discussion.

**Comment by Ladicek** [comment]
Right you are. I did a little bit of digging and the first change you suggested is actually enough to make the test pass with our implementation.

**Comment by Ladicek** [comment]
I'll submit a PR.

**Comment by Ladicek** [comment]
I forgot to mention the issue number in the PR, but this has been resolved in #410.
